### PR TITLE
linter: ignore whitespace at EOL

### PR DESCRIPTION
**Pull request by gares**
This commit contains one of these, so to tests the flag

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Why not remove the `trailing-space` setting in `.gitattributes` instead? (might need to be negated instead of removed since it's a git default)

**Comment by SkySkimmer** [comment]
Going through gitattributes forces `rebase --whitespace=fix` to behave the same as the linter. Not sure if I prefer it being more powerful or it being the same.

**Comment by Zimmi48** [comment]
Can you provide some motivation to remove this check? Is it so annoying?

**Comment by Zimmi48** [comment]
BTW what about testing Gaetan's git commit hook instead of making the linter weaker?

**Comment by gares** [comment]
The motivation is that, as far as I can tell from the previous discussions on the linter, nobody came up with a good reason for signalling this class of errors. The last message in the PR moving the linter to allowed-failures is by @Zimmi48 suggesting that maybe switching off this specific check was a better solution. Finally, I still see PR failing because of this, getting a need:fix label... this is simply frustrating, and given that nobody fond a good reason for this check (other than "it was the default"), here my PR.

This PR is trivial, other "better" proposal in development should not prevent this from being merged IMHO.

**Comment by gares** [comment]
FTR, I have no PR of mine being hit by the linter.

**Comment by Zimmi48** [comment]
I'm pretty convinced by your explanation. I will add that the linter risks to make the life of new contributors harder and making it harder for seemingly no good reason is pretty bad.
Letting the `.gitattributes` file unchanged and encouraging developers to use the commit hook can indeed help ensure that most commits are still free of trailing whitespace without making this a hard requirement.

**Comment by JasonGross** [comment]
Is there a way to set things up so that trailing whitespaces are stripped on merging PRs?  Alternatively, can we set up `./configure` to ask if you want a `.git` linter hook inserted if we're in a git repository?  I'd like to keep the property that commits don't introduce trailing whitespace, and if we can set this up to be mostly invisible to new contributors, that'd be nice.

**Comment by gares** [comment]
Why, what is the problem with trailing spaces?
 
Anyway, please open a PR on configure.ml that installs the whitespace trimmer! It is the right thing to do from a human-machine interaction POV: a machine cannot ask me to fix something that i do not see, it should rather fit it for me, if needed.

**Comment by gares** [comment]
Failures seem to indicate that the overlay system is broken: I did use CI* instead of TRAVIS* in the API pr, and now Travis ignores the overlay. @skyskimmer

File: dev/lint-commits.sh [hunk 1]
```diff
@@ -20,8 +20,8 @@ BASE_COMMIT="$1"
 HEAD_COMMIT="$2"
 
 # git diff --check
-# uses .gitattributes to know what to check
-if git diff --check "$BASE_COMMIT" "$HEAD_COMMIT";
+# uses .gitattributes to know what to check 
+if git diff --check --ignore-space-at-eol "$BASE_COMMIT" "$HEAD_COMMIT";
```
Comment (SkySkimmer) [inline dev/lint-commits.sh]: You should probably change the message at line 29 btw.

**Comment by SkySkimmer** [comment]
>Failures seem to indicate that the overlay system is broken: I did use CI* instead of TRAVIS* in the API pr, and now Travis ignores the overlay. @SkySkimmer

There is no overlay in this PR.

**Comment by SkySkimmer** [comment]
Testing locally shows the `--ignore-space-at-eol` flag doesn't disable the check.
- simple solution: do the gitattributes change, giving up on fixing trailing space automatically
- more complicated: modifying gitattributes in place for the travis job won't work because `diff --check` uses the attributes at the time of the commit, but maybe we could rebase on top of a commit doing the modification then check? 😐 
- even more complicated: the `git diff --check` output is a list of line pairs (error location + type, error line), we could in principle postprocess that to remove the `trailing whitespace` errors

**Comment by SkySkimmer** [comment]
>Is there a way to set things up so that trailing whitespaces are stripped on merging PRs? 

It's certainly possible to stop before committing the merge in order to add changes, which we could make a `fix whitespace for PR #foo` commit for or just include in the merge commit. I don't know how easy it would be to automate.

**Comment by ejgallego** [comment]
Well, I think that if we remove this, we can get rid of the linter as well.

Whitespace at EOL is annoying and useless, I fail to see why we should allow such a sloppy programming pracitce. Fix your toolchain so that newly produced code is sane and doesn't contain useless stuff.

As for new contributors, again I don't see where the problem is in waiving the linter if the circumstances are needed, otherwise they are just an easy command away from producing a correct diff. 

NACK from me.

**Comment by SkySkimmer** [comment]
>Well, I think that if we remove this, we can get rid of the linter as well.

The linter was introduced because tabs for indentation interfered with automation such as ocp-indent. That has nothing to do with trailing space.

**Comment by gares** [comment]
@skyskimmer i did test It locally and It worked, so i really don't know what goes wrong 😞

@ejgallego repeating white spaces are annoying without saying why is not helpful.

**Comment by SkySkimmer** [comment]
To be clear, if you do `git diff --check --ignore-space-at-eol 2d6e395dead61a49ede6208bc40e16b4b8e68ce4 c5db630597589d953531bb267f09e2a10139d2a0` it prints no errors and exits with `0`?

**Comment by gares** [comment]
Im in a pagine now, iirc i did test It in the git index, not a commit

**Comment by ejgallego** [comment]
> @ejgallego repeating white spaces are annoying without saying why is not helpful.

I fail to understand why they are useful either. What is the reason for a coder to add empty spaces at the end of the line? Should it be OK if I add 90 blank spaces at EOL?

I tend to see indentation a the global distribution of whitespace and indeed whitespace at EOL are annoying, and difficult editing from time to time.

For example, take the typical edit operation of making two lines into one, if whitespace at EOL is present, I need to remove the extra spaces instead of just pressing back in my editor.

**Comment by Zimmi48** [comment]
The idea of @JasonGross to install the git commit hook at configure time looks good to me. Maybe this should be implemented in #6283. It could test that the commit hook is already there (checking that this is the exact same commit hook), then propose to install it / override it.

If this makes the linter painless, then it's not necessary to make it weaker, especially given that some developers still seem to think that trailing white spaces are an annoyance.

**Comment by gares** [comment]
> For example, take the typical edit operation of making two lines into one, if whitespace at EOL is present, I need to remove the extra spaces instead of just pressing back in my editor.

This is funny. In my text editor, vim, to join two lines you press `J` and having a trailing whitespace helps, otherwise you have to add it yourself. What should I do? Add a converse linter check and tell people to install vim^W^W "configure their toolchain"?

**Pull request closed without merging**